File: DSA/Akshat_vaid_590011590/Day_23/day23_question2.py

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def build_linked_list(values):
    if not values:
        return None

    head = ListNode(values[0])
    current = head

    for val in values[1:]:
        current.next = ListNode(val)
        current = current.next

    return head


def print_linked_list(head):
    current = head
    while current:
        print(current.val, end=" -> " if current.next else "")
        current = current.next
    print()


def remove_duplicates(head):
    current = head

    while current and current.next:
        if current.val == current.next.val:
            
            current.next = current.next.next
        else:
            
            current = current.next

    return head


values = [1, 1, 2, 3, 3, 4, 4, 4, 5]
head = build_linked_list(values)

print("Original Linked List:")
print_linked_list(head)

head = remove_duplicates(head)

print("After Removing Duplicates:")
print_linked_list(head)

File: DSA/Khushi_Dangwal_500124475/Day_23_question2.c
#include <stdio.h>
#include <stdlib.h>


struct ListNode {
    int val;
    struct ListNode *next;
};

// remove duplicates from sorted list
struct ListNode* removeDuplicates(struct ListNode* head) {
    if (head == NULL) {
        return NULL;
    }
    
    struct ListNode* current = head;
    
    while (current->next != NULL) {
        if (current->val == current->next->val) {
            // Duplicate found - remove the next node
            struct ListNode* temp = current->next;
            current->next = current->next->next;
            free(temp); // Free memory of duplicate node
        } else {
            // No duplicate - move to next node
            current = current->next;
        }
    }
    
    return head;
}

// Helper function to create a new node
struct ListNode* newNode(int val) {
    struct ListNode* node = (struct ListNode*)malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = NULL;
    return node;
}

// Helper function to print the list
void printList(struct ListNode* head) {
    while (head != NULL) {
        printf("%d", head->val);
        if (head->next != NULL) {
            printf(" -> ");
        }
        head = head->next;
    }
    printf("\n");
}

// Main function to test the code
int main() {
    // Create example linked list: 2->2->4->5
    struct ListNode* head = newNode(2);
    head->next = newNode(2);
    head->next->next = newNode(4);
    head->next->next->next = newNode(5);
    
    printf("Original list: ");
    printList(head);
    
    head = removeDuplicates(head);
    printf("List after removing duplicates: ");
    printList(head);
    
    // Create another example: 2->2->2->2->2
    struct ListNode* head2 = newNode(2);
    head2->next = newNode(2);
    head2->next->next = newNode(2);
    head2->next->next->next = newNode(2);
    head2->next->next->next->next = newNode(2);
    
    printf("\nOriginal list: ");
    printList(head2);
    
    head2 = removeDuplicates(head2);
    printf("List after removing duplicates: ");
    printList(head2);
    
    return 0;
}
